kong/router/transform.lua

```diff
@@ -690,48 +689,59 @@ local function group_by(t, f)
-  local original_route = route_and_service.route
+-- split routes into multiple routes,
+-- one for each prefix length and one for all regular expressions
+local function split_routes_and_services_by_path(routes_and_services)
+  local count = #routes_and_services
+  local append_count = 1
 
-  if is_empty_field(original_route.paths) or #original_route.paths == 1 or
-     not is_null(original_route.expression) -- expression will ignore paths
-  then
-    tb_insert(routes_and_services_split, route_and_service)
-    return
-  end
+  for i = 1, count do
+    local route_and_service = routes_and_services[i]
+    local original_route = route_and_service.route
+    local original_paths = original_route.paths
 
-  -- make sure that route_and_service contains only the two expected entries, route and service
-  assert(tb_nkeys(route_and_service) == 1 or tb_nkeys(route_and_service) == 2)
+    if is_empty_field(original_paths) or #original_paths == 1 or
+       not is_null(original_route.expression) -- expression will ignore paths
+    then
+      goto continue
+    end
 
-  local grouped_paths = group_by(
-    original_route.paths, sort_by_regex_or_length
-  )
-  for index, paths in pairs(grouped_paths) do
-    local cloned_route = {
-      route = shallow_copy(original_route),
-      service = route_and_service.service,
-    }
+    -- make sure that route_and_service contains only the two expected entries, route and service
+    assert(tb_nkeys(route_and_service) == 1 or tb_nkeys(route_and_service) == 2)
 
-    cloned_route.route.original_route = original_route
-    cloned_route.route.paths = paths
-    cloned_route.route.id = uuid_generator(original_route.id .. "#" .. tostring(index))
+    local original_route_id = original_route.id
+    local original_service = route_and_service.service
+    local grouped_paths = group_by(original_paths, sort_by_regex_or_length)
 
-    tb_insert(routes_and_services_split, cloned_route)
-  end
-end
+    local is_first = true
+    for idx, paths in pairs(grouped_paths) do
+      local route = shallow_copy(original_route)
 
+      route.original_route = original_route
+      route.paths = paths
+      route.id = uuid_generator(original_route_id .. "#" .. tostring(idx))
 
-local function split_routes_and_services_by_path(routes_and_services)
-  local count = #routes_and_services
-  local routes_and_services_split = tb_new(count, 0)
+      local cloned_route_and_service = {
+        route = route,
+        service = original_service,
+      }
 
-  for i = 1, count do
-    split_route_by_path_info(routes_and_services[i], routes_and_services_split)
-  end
+      if is_first then
+        -- the first one will replace the original route
+        routes_and_services[i] = cloned_route_and_service
+        is_first = false
+
+      else
+        -- the others will append to the original routes array
+        routes_and_services[count + append_count] = cloned_route_and_service
```

Comment:
And just do it here:

```suggestion
        routes_and_services[count + append_count] = {
        route = route,
        service = original_service,
      }
```

Reply:
As Hans said, this logic is a little hard to understand, I also spent many time to figure it out then refactor it.